File: gosource/utils/array.ts

/**
 * Returns a new array with only the last occurrence of each unique item kept.
 * @param array Input array with possible duplicates.
 */
function setback<T>(array: T[]): T[] {
	const lastSeen = new Map<T, number>();
	array.forEach((item, index) => lastSeen.set(item, index));

	return array.filter((item, index) => lastSeen.get(item) === index);
}

/**
 * Converts a numbered object into an array up to `maxKey`.
 * Missing entries are replaced with empty arrays.
 * @param obj Number-indexed object.
 * @param maxKey Maximum index to include.
 */
function fromNumberedObject<T>(obj: Record<number, T[]>, maxKey: number): T[][] {
	return Array.from({ length: maxKey + 1 }, (_, i) => obj[i] ?? []);
}

/**
 * Finds the longest sub-chain from `child` that appears in `parent` in the same order.
 * @param parent Reference sequence.
 * @param child Subsequence to check.
 */
function longestSubChain<T>(parent: T[] = [], child: T[] = []): T[] {
	if (parent.length === 0 || child.length === 0) { return []; }

	const results: T[][] = [];
	let remainingChild = [...child];
	let maxScore = 0;
	let resultIndex = 0;
	let parentInNow = 0;
	let parentInLast = 0;

	while (remainingChild.length) {
		parentInLast = -1;
		const currentChain: T[] = [];
		const remainingChildNext: T[] = [];

		for (
			let index = child.indexOf(remainingChild[0]);
			index < child.length;
			index++
		) {
			parentInNow = parent.indexOf(child[index]);

			if (parentInLast < parentInNow) {
				currentChain.push(child[index]);
				parentInLast = parentInNow;
			} else if (
				remainingChild.includes(child[index]) &&
				parent.includes(child[index])
			) {
				remainingChildNext.push(child[index]);
			}
		}

		if (maxScore < currentChain.length) {
			maxScore = currentChain.length;
			resultIndex = results.length;
			results.push(currentChain);
		}

		remainingChild = remainingChildNext;
	}

	return results[resultIndex] ?? [];
}


/**
 * Check if subseq is a subsequence of sequence (maintains order)
 */
function isSubsequence(subseq: number[], sequence: number[]): boolean {
	if (subseq.length === 0) {
		return true;
	}

	let subseqIndex = 0;

	for (const element of sequence) {
		if (subseqIndex < subseq.length && element === subseq[subseqIndex]) {
			subseqIndex++;
			if (subseqIndex === subseq.length) {
				return true;
			}
		}
	}

	return subseqIndex === subseq.length;
}

function findArrSuperParent(array: number[], findFromArrays: number[][]): number[] | null {
	for (const candidate of findFromArrays) {
		if (isSubsequence(array, candidate)) {
			return candidate;
		}
	}

	return null;
}

export default {
	setback,
	fromNumberedObject,
	longestSubChain,
	isSubsequence,
	findArrSuperParent
};

File: internal/order/array.go
package order

import "slices"

func Array_LongestSubChain[T comparable](parent, child []T) []T {
	if len(parent) == 0 || len(child) == 0 {
		return nil
	}
	results := [][]T{}
	remainingChild := append([]T{}, child...)
	maxScore := 0
	resultIndex := 0

	for len(remainingChild) > 0 {
		parentInLast := -1
		currentChain := []T{}
		remainingChildNext := []T{}

		startIndex := 0
		for i, v := range child {
			if v == remainingChild[0] {
				startIndex = i
				break
			}
		}
		for index := startIndex; index < len(child); index++ {
			parentInNow := -1
			for j, v := range parent {
				if v == child[index] {
					parentInNow = j
					break
				}
			}
			if parentInNow > parentInLast {
				currentChain = append(currentChain, child[index])
				parentInLast = parentInNow
			} else {
				if slices.Contains(remainingChild, child[index]) {
					remainingChildNext = append(remainingChildNext, child[index])
				}
			}
		}
		if len(currentChain) > maxScore {
			maxScore = len(currentChain)
			resultIndex = len(results)
			results = append(results, currentChain)
		}
		remainingChild = remainingChildNext
	}
	if resultIndex < len(results) {
		return results[resultIndex]
	}
	return nil
}

func Array_IsSubsequence(subseq, sequence []int) bool {
	if len(subseq) == 0 {
		return true
	}
	subseqIndex := 0
	for _, element := range sequence {
		if subseqIndex < len(subseq) && element == subseq[subseqIndex] {
			subseqIndex++
			if subseqIndex == len(subseq) {
				return true
			}
		}
	}
	return subseqIndex == len(subseq)
}

func Array_FindSuperParent(array []int, findFromArrays [][]int) []int {
	for _, candidate := range findFromArrays {
		if Array_IsSubsequence(array, candidate) {
			return candidate
		}
	}
	return nil
}
